src/py/mat3ra/made/tools/build/passivation/builders.py

```diff
@@ -0,0 +1,215 @@
+            material=material,
+            surface=SurfaceTypes.TOP,
+            shadowing_radius=self.build_parameters.shadowing_radius,
+            depth=self.build_parameters.depth,
+        )
+        undercoordinated_atoms_indices = get_undercoordinated_atom_indices(
+            material=material,
+            indices=surface_atoms_indices,
+            cutoff=self.build_parameters.shadowing_radius,
+            coordination_threshold=self.build_parameters.coordination_threshold,
+        )
+        passivant_coordinates_values = self._get_passivant_coordinates(
+            material, configuration, undercoordinated_atoms_indices
+        )
+        return self._add_passivant_atoms(material, passivant_coordinates_values, configuration.passivant)
+
+    def _get_passivant_coordinates(
+        self, material: Material, configuration: PassivationConfiguration, undercoordinated_atoms_indices: list
+    ):
+        """
+        Calculate the coordinates for placing passivating atoms based on the specified edge type.
+
+        Args:
+            material (Material): Material to passivate.
+            configuration (SurfacePassivationConfiguration): Configuration for passivation.
+            undercoordinated_atoms_indices (list): Indices of undercoordinated atoms.
+        """
+        passivant_coordinates = []
+        for idx in undercoordinated_atoms_indices:
+            nearest_neighbors = get_nearest_neighbors_atom_indices(
+                material=material,
+                coordinate=material.basis.coordinates.get_element_value_by_index(idx),
+                cutoff=self.build_parameters.shadowing_radius,
+            )
+            if nearest_neighbors is None:
+                continue
+            average_coordinate = np.mean(
+                [material.basis.coordinates.get_element_value_by_index(i) for i in nearest_neighbors], axis=0
+            )
+            bond_vector = material.basis.coordinates.get_element_value_by_index(idx) - average_coordinate
+            bond_vector = bond_vector / np.linalg.norm(bond_vector) * configuration.bond_length
+            passivant_bond_vector_crystal = material.basis.cell.convert_point_to_crystal(bond_vector)
+
+            passivant_coordinates.append(
+                material.basis.coordinates.get_element_value_by_index(idx) + passivant_bond_vector_crystal
+            )
+
+        return passivant_coordinates
+
+    def get_coordination_numbers(self, material: Material):
+        """
+        Get the coordination numbers for all atoms in the material.
+
+        Args:
+            material (Material): The material object.
+
+        Returns:
+            set: The coordination numbers for all atoms in the material.
+        """
+
+        coordination_numbers = set(
+            get_coordination_numbers(material=material, cutoff=self.build_parameters.shadowing_radius)
+        )
+        print("coordination numbers:", coordination_numbers)
```

Comment:
1. Let's name this function `get_unique_coordination_numbers`
2. Update the description to state accordingly
3. Remove the print statement